skimage/segmentation/random_walker_segmentation.py

```diff
@@ -491,19 +493,37 @@ def random_walker(
             f"'cg', 'cg_j', 'bf', and None"
         )
 
+    if data.dtype == np.float16:
+        # SciPy sparse, which is used later on, doesn't officially support float16
+        # This led to failures when testing with NumPy 1.26 (see gh-7635).
+        warnings.warn(
+            message="upcasting `data` with dtype float16 to float32",
+            stacklevel=3,
+        )
+        data = data.astype(np.float32, casting="safe")
+
     # Spacing kwarg checks
     if spacing is None:
-        spacing = np.ones(3)
-    elif len(spacing) == labels.ndim:
+        spacing = np.ones(3, dtype=data.dtype)
+    elif not isinstance(spacing, np.ndarray):
+        spacing = np.asarray(spacing, dtype=data.dtype)
```

Comment:
I'm not sure if all this input validation is a good call. Since spacing can be an iterable, we need to make sure that get's converted to an array with correct dtype. 

But I think we can't use `asarray` if `spacing` is already an array since that doesn't appear to use safe casting: e.g. `np.asarray(np.array([200, 200]), dtype=np.int8)` happily overflows. So we need to use `astype(..., casting="safe") explicitly below. :thinking: 

Am I confusing myself? Am I missing an easy solution here? This kind of things keeps coming up and I'm starting to think tha tasking for more restrictive types might be a good idea. It's convenient for users, but is that worth all the input validation logic? It seems very easy to miss edge cases, especially between different NumPy versions... :see_no_evil:

Reply:
From https://github.com/scikit-image/scikit-image/pull/7655/commits/ec8d0a363846a11cc7effb3af8d1f2d5afe92dff:

> We decided that trying to preserve float32 isn't worth the effort; there
wasn't really a guarantee that we did so previously. We just make sure
to upcast from float16 which might lead to failures.

Reply:
@lagru sorry for being off-topic and/or late to the party, but I'm just curious:

> e.g. `np.asarray(np.array([200, 200]), dtype=np.int8)` happily overflows

for which versions of NumPy and/or Python?

Reply:
numpy 2 with recent python versions....  
Are there other versions that don't give the output:
``` array([-56, -56], dtype=int8)```

Reply:
:+1: I haven't checked earlier NumPy versions.

Reply:
np1.23, 1.24, 2.0, 2.2 on python 3.10, python 3.13. All show `asarray` silently overflowing.

Reply:
An option for safer integer casts would be nice.  But until then, yes: converting python integers has stricter rules.